.github/workflows/test.yml

```diff
@@ -0,0 +1,292 @@
+      - name: Cache cargo index
+        uses: actions/cache@v1.1.2
+        with:
+          path: ~/.cargo/git
+          key: ${{ runner.os }}-cargo-git-${{ hashFiles('**/Cargo.lock') }}
+      - name: Cache cargo build
+        uses: actions/cache@v1.1.2
+        with:
+          path: target
+          key: ${{ runner.os }}-cargo-build-target-${{ hashFiles('**/Cargo.lock') }}
+      # Install sccache based on https://github.com/denoland/rusty_v8/blob/683aa2b2a981c7991932b793a8203ec245649d7a/.github/workflows/ci.yml#L43
+      - name: Install sccache for ${{ matrix.platform}}
+        env:
+          SCCACHE_CACHE_SIZE: "1G"
+          SCCACHE_IDLE_TIMEOUT: 0
+        shell: pwsh
+        run: |
+          $version = "0.2.12"
+          $platform =
+            @{ "macOS"   = "x86_64-apple-darwin"
+               "Linux"   = "x86_64-unknown-linux-musl"
+               "Windows" = "x86_64-pc-windows-msvc"
+             }.${{ runner.os }}
+          $basename = "sccache-$version-$platform"
+          $url = "https://github.com/mozilla/sccache/releases/download/" +
+                 "$version/$basename.tar.gz"
+          curl -LO $url
+          tar -xzvf "$basename.tar.gz"
+          . $basename/sccache --start-server
+          echo "::add-path::$(pwd)/$basename"
+          echo "::set-env name=RUSTC_WRAPPER::sccache"
+      - name: Cache sccache
+        uses: actions/cache@v1.1.2
+        with:
+          path: ~/.cache/sccache
+          key: ${{ runner.os }}-sccache-check-${{ hashFiles('**/Cargo.lock') }}
+      - name: Run cargo check 1/3
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --locked --no-default-features --verbose
+      - name: Run cargo check 2/3
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --locked --manifest-path util/io/Cargo.toml --no-default-features --verbose
+      - name: Run cargo check 3/3
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --locked --manifest-path util/io/Cargo.toml --features "mio" --verbose
+      - name: Run cargo check evmbin
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --locked -p evmbin --verbose
+      - name: Run cargo check benches
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --locked --all --benches --verbose
+      - name: Run validate chainspecs
+        run: ./scripts/actions/validate-chainspecs.sh
+      - name: Sccache statistics
+        run: sccache --show-stats
+      # From https://github.com/launchbadge/sqlx/blob/master/.github/workflows/rust.yml
+      - name: Prepare build directory for cache
+        run: |
+          find ./target/debug -maxdepth 1 -type f -delete \
+          && rm -fr ./target/debug/{deps,.fingerprint}/*{parity,ethcore,ethkey,ethstore,parity-evm}* \
+          && rm -f  ./target/.rustc_info.json
+  build-tests:
```

Comment:
The reason I had build-tests separate from run-tests step is to easily measure how long is each step and what can be reused after build step is cached. I don't think we need this separation anymore, it only slows down the tests.
As you can see in https://github.com/OpenEthereum/open-ethereum/pull/11528/checks?check_run_id=474452118, macOS test builds are faster than on linux (as noted in https://github.com/ordian/open-ethereum-ci-experiment/issues/7#issue-554210166), this needs to be investigated as well as whether we need both sccache and actions/cache.

Reply:
As you can see in https://github.com/OpenEthereum/open-ethereum/runs/474519539#step:13:689 run-tests recompilation on macOS took 15 mins, so the separation of running the tests vs build them only slows down the pipeline. Of course that's only a symptom of caching not working properly, but it may take a long time till we fix it.

Reply:
https://github.com/actions/cache/issues/133#issuecomment-592476433